### ISSUE TITLE
Formalize PluginAdapter semantics

### ISSUE THREAD

**Issue by decentralion** [sourcecred/sourcecred#465]
- Can two plugins have the same prefix?
- Can two plugins have the same display name?
- Can one plugin have two types with the same prefix?
- Can one plugin have a prefix that is a prefix of another plugin's prefix?
- Can two plugins have types with the same prefix?

Tune in next episode for answers (and invariant assertions) to all these questions, and more!

**Comment by wchargin** [comment]
Cool; I like this. I was going to take a stab at implementing this, but
it sounds like you have a lot of context/prior work on it, so I’ll hold
off on doing so.

> They both have names

Note that names of plugins could _arguably_ be required to be unique
(though I wouldn’t be thrilled about this), but names of types are
definitely not unique. Concepts of “user”, “issue”, “comment” will be
shared across lots of plugins. (No change required.)

> What if we unify them into a shared recursive abstraction

I don’t see how the abstraction that you propose is recursive. What do
you mean by that?

> Then, when finding the right method for a given node address, we choose the most specific Type available (and the presence of the BaseNodeType ensures that we always find a type).

I like the idea of having the base node type being a “fallback” on the
same level of everything else.

> It is illegal to have multiple types with the exact same prefix.

Yes, agreed. Perhaps we should also require that if `t1` is a child type
of `t0` (e.g., `t1` is a node type and `t0` is a plugin), then the
prefix of `t0` is a prefix of the prefix of `t1`. (That is, the function
from node types to node prefixes is a poset homomorphism.) Then, it is
possible to ensure that all types have distinct prefixes by reasoning
locally: you only have to consider your siblings, not all other types at
all levels, and in particular it will be valid to combine any set of
plugins as long as their _top-level_ prefixes are pairwise disjoint.

> `type NodeType`

Multiple functions of the `NodeType` (`render`, `weight`, at least) will
need to take more context than just a single node address—the graph, at
the very least, plus maybe some additional context (e.g., the GitHub
plugin’s relational view). But it sounds like you’re suggesting that the
node types be constants, as opposed to (e.g.) “creating” the GitHub node
types for a particular repository owner and name, receiving a
`Promise<NodeType[]>` in return. How is this resolved?

> As written, the types have all kinds of stuff attached (render,
> weight, etc), but clearly we should factor that out.

Agreed, though I’m willing to do this as needed.

**Comment by decentralion** [comment]
Another idea, based on some discussions with @wchargin:

Often times, consumers of the static adapter actually want the data in this type:
```
export type PluginDescription = {|
  +nodePrefix: NodeAddressT,
  +edgePrefix: EdgeAddressT,
  +name: string,
  +nodeTypes: $ReadOnlyArray<NodeType>,
  +edgeTypes: $ReadOnlyArray<EdgeType>,
|}
```

Interestingly, this is almost exactly the static adapter, minus the `load` method. As such, I now think it would be nice to refactor the adapters into two classes: `UnloadedAdapter` and `LoadedAdapter`.
The loaded adapter would provide dynamically available methods like the `nodeDescription`, `graph`, - basically it absorbs the DynamicAdapter. It would also have a method to retrieve the `pluginDescription` type defined above, and an `unload` method in case you need an UnloadedAdapter. The `UnloadedAdapter` would provide a method to get the `pluginDescription` data type above, and provide a `load`method for getting the `LoadedAdapter`.

Basically, this factors away the parts of the static adapter that just provide data, from the part that is responsible for loading a dynamic adapter. This is convenient as many consumers of the static adapter's data don't need the extra complexity of being able to load the adapter, and having that complexity around makes it less convenient to test consumers of the static adapter.

Also, moving away from an interface to a simple data type will be convenient.

Will need to make a corresponding refactor to the adapter sets, so as to create a PluginDescriptionSet or such, with intelligent use of Tries.

**Comment by wchargin** [comment]
Thanks for writing this up. The plan makes sense to me.

> Also, moving away from an interface to a simple data type will be
> convenient.

Citation needed; I don’t see what you mean.

> This is convenient as many consumers of the static adapter's data
> don't need the extra complexity of being able to load the adapter, and
> having that complexity around makes it less convenient to test
> consumers of the static adapter.

Yep. As mentioned offline, the ability to load the dynamic adapter also
incurs a lot of power that we don’t necessarily want to grant to (e.g.)
the weight config UI.

> Will need to make a corresponding refactor to the adapter sets, so as
> to create a PluginDescriptionSet or such, with intelligent use of
> Tries.

Agreed. This suggests that it might be nice, for consistency, to let
`PluginDescription`/`StaticPluginAdapter` be a class. Or maybe not; it
doesn’t seem to have huge concrete advantages.
